enumerables.rb

```diff
@@ -0,0 +1,101 @@
+module Enumerable
+  def my_each
+    return enum_for unless block_given?
+
+    elements = is_a?(Range) ? to_a : self
+    index = 0
+    elements.size.times do
+      yield(elements[index])
+      index += 1
+    end
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    elements = is_a?(Range) ? to_a : self
+    index = 0
+    size.times do
+      yield(elements[index], index)
+      index += 1
+    end
+  end
+
+  def my_select(&array)
+    return enum_for unless block_given?
+
+    selected_items = []
+    my_each do |x|
+      selected_items.push(x) unless array.call(x) == false
+    end
+    selected_items
+  end
+
+  def my_all(*args)
+    my_each do |x|
+      return true unless block_given?
+
+      return false unless yield(x)
+    end
+    true
+  end
+
+  def my_any?(*args)
+    args = nil
+    if args
+      my_each { |x| return false unless x == args }
+    end
+    if !block_given?
+      my_each { |x| return true if x }
+    else
+      my_each { |x| return true if x == yield(x) }
+    end
+    false
+  end
+
+  def my_none(*args)
+    !my_any?
+  end
+
+  def my_count(*arguement)
+    counter = 0
+    arguement = nil
+    if arguement
+      my_each { |x| counter += 1 if x == arguement }
+    elsif block_given?
+      my_each { |x| counter += 1 if yield x }
+    else
+      counter = length
+    end
+    counter
+  end
+
+  def my_map(proc = nil)
+    return to_enum unless block_given?
+
+    selected = []
+    if proc
+      my_each { |x| selected.push(proc.call(x)) }
+    else
+      my_each { |x| selected.push(yield(x)) }
+    end
+    selected
+  end
+
+  def my_inject(*args, &block)
+    element = nil
+    symbol = nil
+    element = element.to_sym if element.is_a?(String) && !Symbol && !Block
+    if element.is_a?(Symbol) && !symbol
+      block = element.to_proc
+      element = nil
+    end
+    symbol = symbol.to_sym if symbol.is_a?(String)
+    block = symbol.to_proc if symbol.is_a?(Symbol)
+    my_each { |x| element = element.nil? ? x : block.yield(element, x) }
+    element
+  end
```

Comment:
* [ ] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block
  ```ruby
    array = [4,7,8,89,9,6,5,12,4,2,5]
    operation = proc { |sum, n| sum + n  }
    p array.my_inject(&operation) == array.inject(&operation) # false
  ```

* [ ] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method
  ```ruby
    array = [1,7,8,89,9,6,5,12,4,2,5]
    p array.my_inject(:+) == array.inject(:+) # return this error
    # NoMethodError:
        undefined method `yield' for nil:NilClass
      # ./enumerables.rb:95:in `block in my_inject'
      # ./enumerables.rb:8:in `block in my_each'
      # ./enumerables.rb:7:in `times'
      # ./enumerables.rb:7:in `my_each'
      # ./enumerables.rb:95:in `my_inject'
  ```